src/Workspaces/Remote/Core/AbstractAssetProviderExtensions.AnalyzerReference.cs

```diff
@@ -0,0 +1,149 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Serialization;
+using Roslyn.Utilities;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+#if NET
+using System.Runtime.Loader;
+#endif
+
+namespace Microsoft.CodeAnalysis.Remote;
+
+internal static partial class AbstractAssetProviderExtensions
+{
+#if NET
+
+    /// <summary>
+    /// Gate around <see cref="s_checksumToReferenceSet"/> to ensure it is only accessed and updated atomically.
+    /// </summary>
+    private static readonly SemaphoreSlim s_isolatedReferenceSetGate = new(initialCount: 1);
+
+    /// <summary>
+    /// Mapping from checksum for a particular set of assembly references, to the dedicated ALC and actual assembly
+    /// references corresponding to it.  As long as it is alive, we will try to reuse what is in memory.  But once it is
+    /// dropped from memory, we'll clean things up and produce a new one.
+    /// </summary>
+    private static readonly Dictionary<Checksum, WeakReference<IsolatedAssemblyReferenceSet>> s_checksumToReferenceSet = [];
+
+    private static void GarbageCollectReleaseReferences_NoLock()
+    {
+        Contract.ThrowIfTrue(s_isolatedReferenceSetGate.CurrentCount != 0);
+
+        // Any time we've grown the dictionary by a substantive amount, do a sweep to see if we can remove any entries.
+        if (s_checksumToReferenceSet.Count % 128 != 0)
```

Comment:
>if (s_checksumToReferenceSet.Count % 128 != 0) [](http://example.com/codeflow?start=8&length=46)

Playing devil's advocate: What if toggle between 1279 and 1280 entries repeatedly? Seems like we could spend a lot of time in here. Would it make more sense to just do this every 100 or so times this method is entered?